rtmr/rtmr.go

```diff
@@ -0,0 +1,208 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rtmr
+
+import (
+	"crypto"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/google/go-configfs-tsm/configfs/configfsi"
+)
+
+const (
+	RtmrSubsystem = "rtmrs"
+	TsmRtmrPrefix = configfsi.TsmPrefix + "/" + RtmrSubsystem
+	TsmRtmrDigest = "digest"
+	TsmPathIndex  = "index"
+	TsmPathTcgMap = "tcg_map"
+)
+
+// RtmrExtend is a struct that represents a rtmr entry in the configfs.
+type RtmrExtend struct {
+	RtmrIndex int
+	entry     *configfsi.TsmPath
+	client    configfsi.Client
+}
+
+type RtmrResponse struct {
+	RtmrIndex int
+	digest    []byte
+	tcgMap    []byte
+}
+
+func (r *RtmrExtend) attribute(subtree string) string {
+	a := *r.entry
+	a.Attribute = subtree
+	return a.String()
+}
+
+// extendDigest extends the measurement to the rtmr with the given hash.
+func (r *RtmrExtend) extendDigest(hash []byte) error {
+	if err := r.client.WriteFile(r.attribute(TsmRtmrDigest), hash); err != nil {
+		return fmt.Errorf("could not write digest to rmtr%d: %v", r.RtmrIndex, err)
+	}
+	return nil
+}
+
+// getDigest returns the digest of the rtmr.
+func (r *RtmrExtend) getDigest() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(TsmRtmrDigest))
+}
+
+// getTcgMap returns the tcg map of the rtmr.
+func (r *RtmrExtend) getTcgMap() ([]byte, error) {
+	return r.client.ReadFile(r.attribute(TsmPathTcgMap))
+}
+
+// validateRtmrIndex checks if the rtmr index is valid.
+func (r *RtmrExtend) validateRtmrIndex() error {
+	indexBytes, err := r.client.ReadFile(r.attribute(TsmPathIndex))
+	if err != nil {
+		return fmt.Errorf("could not read index %s: %v", r.attribute(TsmPathIndex), err)
+	}
+	index, err := configfsi.Kstrtouint(indexBytes, 10, 64)
+	if err != nil {
+		return fmt.Errorf("could not convert index %s to integer: %v", indexBytes, err)
+	}
+	if (int)(index) != r.RtmrIndex {
+		return fmt.Errorf("rtmr index %d does not match the expected index %d", index, r.RtmrIndex)
+	}
+	return nil
+}
+
+// setRtmrIndex sets a configfs rtmr entry to the given index.
+// It reports an error if the index cannot be written.
+func (r *RtmrExtend) setRtmrIndex() error {
+	indexBytes := []byte(strconv.Itoa(r.RtmrIndex)) // Convert index to []byte
+	indexPath := r.attribute(TsmPathIndex)
+	if err := r.client.WriteFile(indexPath, indexBytes); err != nil {
+		return fmt.Errorf("could not write index %s: %v", indexPath, err)
+	}
+	return nil
+}
+
+// searchRtmrInterface searches for an rtmr entry in the configfs.
+func searchRtmrInterface(client configfsi.Client, index int) (*RtmrExtend, error) {
+	root := TsmRtmrPrefix
+	out := RtmrExtend{}
+	err := filepath.WalkDir(root, func(path string, d os.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+		if !d.IsDir() {
+			p, err := configfsi.ParseTsmPath(path)
+			if err != nil {
+				return err
+			}
+			r := RtmrExtend{
```

Comment:
This should be &RtmrExtend since you call a pointer-receiving function on it after.